### PR TITLE
Integrating support for propagating state to member resources within `ChangeSetPersister::PropagateVisibilityAndState`

### LINKED ISSUE #1240: Refactor ephemera folder publication
Propose: Instead of logic which treats folders as published either if it's complete or its parent is complete, actually propagate the workflow state down to child objects. This would allow us to simplify indexing and ability logic, and also unblock https://github.com/pulibrary/figgy/issues/565

### PR DESCRIPTION
**Pull request by jrgriffiniii**
Resolves #1240

### CODE REVIEW

File: spec/change_sets/scanned_resource_change_set_spec.rb [hunk 1]
```diff
@@ -98,4 +98,20 @@
       expect(change_set.logical_structure[0].label).to eq ["Logical"]
     end
   end
+  context "when an EphemeraBox has EphemeraFolder members" do
```
Comment (eliotjordan) [inline spec/change_sets/scanned_resource_change_set_spec.rb]: Should this be "when a ScannedResource has Scanned Resource members"?

File: app/models/folder_workflow.rb [hunk 2]
```diff
@@ -22,12 +16,25 @@ def initialize(state)
...
+        received: :needs_qa,
+        all_in_production: :complete
+      }
+    }
   end
 
-  def valid_transitions
-    aasm.states(permitted: true).map(&:name).map(&:to_s)
+  # Retrieve the state for the resource
+  # @param related_resource [Valhalla::Resource] the resource related to the current workflow
```
Comment (eliotjordan) [inline app/models/folder_workflow.rb]: Parameter is the resource class not the resource itself.

**Comment by jrgriffiniii** [comment]
I feel that I've addressed some of my concerns by decoupling Workflow state translations from Resource Classes (I overlooked that).

**Comment by jrgriffiniii** [comment]
a9058e2 attempts to implement a simplified approach proposed by @tpendragon when discussing how best to refactor this further.

File: valhalla/app/change_sets/valhalla/change_set.rb [hunk 3]
```diff
@@ -7,6 +7,13 @@ def self.apply_workflow(workflow)
       include(Valhalla::ChangeSetWorkflow)
     end
 
+    def state_for_related(related_resource, related_state)
```
Comment (hackmastera) [inline valhalla/app/change_sets/valhalla/change_set.rb]: where is this used?

File: app/change_set_persisters/change_set_persister/propagate_visibility_and_state.rb [hunk 4]
```diff
@@ -12,13 +12,37 @@ def run
       return if !change_set.changed?(:visibility) && !change_set.changed?(:state)
       members.each do |member|
```
Comment (hackmastera) [inline app/change_set_persisters/change_set_persister/propagate_visibility_and_state.rb]: can you add a class-level comment explaining how the logic works?

**Comment by jrgriffiniii** [comment]
This currently blocks the resolution for #1279 (please see https://github.com/pulibrary/figgy/tree/issues-1279-jrgriffiniii-folders-migration).  Once this is acceptable for merging, I can rebase that branch and issue a separate Pull Request.

**Pull request merged**